Audio PR4 - Adding support for file list reader

SundarRajan28 opened the pull request.

SundarRajan28 commented on rocAL/include/pipeline/master_graph.h:
```diff
@@ -118,6 +118,7 @@ class MasterGraph {
     void box_iou_matcher(std::vector<float> &anchors, float high_threshold, float low_threshold, bool allow_low_quality_matches);
     void create_randombboxcrop_reader(RandomBBoxCrop_MetaDataReaderType reader_type, RandomBBoxCrop_MetaDataType label_type, bool all_boxes_overlap, bool no_crop, FloatParam *aspect_ratio, bool has_shape, int crop_width, int crop_height, int num_attempts, FloatParam *scaling, int total_num_attempts, int64_t seed = 0);
     const std::pair<ImageNameBatch, pMetaDataBatch> &meta_data();
+    void* get_meta_read_buffer_at(uint i);
```
Rename the function. I can't understand the name.

swetha097 replied:
@shobana-mcw Please comment on this

shobana-mcw replied:
change uint i -> uint index

shobana-mcw replied:
Like discussed before add comment for every new API introduced.

SundarRajan28 commented on rocAL_pybind/amd/rocal/pipeline.py:
```diff
@@ -161,11 +160,11 @@ def get_one_hot_encoded_labels(self, array, device):
+            # torch tensor
                 return b.getOneHotEncodedLabels(self._handle, ctypes.c_void_p(array.data_ptr()), self._num_classes, 1)
+    
+    def get_one_hot_encoded_labels_cupy(self, array):
```
Change cupy to gpu. Cupy may be removed in future so better to use gpu instead

swetha097 replied:
This function explicitly handles CuPy arrays, hence named it so

SundarRajan28 commented on rocAL_pybind/amd/rocal/plugin/generic.py:
```diff
@@ -157,8 +157,13 @@ def __next__(self):
             if self.loader._one_hot_encoding == True:
-                self.loader.get_one_hot_encoded_labels(
-                    self.labels, self.device)
+                #cpu
```
Remove comment as if else is self explanatory

SundarRajan28 commented on rocAL_pybind/amd/rocal/plugin/generic.py:
```diff
@@ -157,8 +157,13 @@ def __next__(self):
-                    self.labels, self.device)
+                #cpu
+                if self.device == "cpu":
+                    self.loader.get_one_hot_encoded_labels(
```
Add _cpu to the function name to differentiate the two variants.

swetha097 replied:
This handles both the CPU and GPU variants for torch counterparts.

The differeciation is not possible , hence added another function.

How are you suggesting to to do the differciation in pipeline.py file?

shobana-mcw replied:
@SundarRajan28 Do you have any other suggestion?

SundarRajan28 commented on rocAL_pybind/rocal_pybind.cpp:
```diff
@@ -129,6 +129,20 @@ py::object wrapperRocalExternalSourceFeedInput(
+        return py::cast<py::none>(Py_None);
+    }
+
+    py::object wrapper_cupy_one_hot_label_copy(RocalContext context, size_t array_ptr, unsigned num_of_classes) {
```
Remove cupy and see if u can combine both functions into a single one while differentiating the cpu/gpu with a func param

swetha097 replied:
I think the args are diff, hence cannot be combined

SundarRajan28 commented on rocAL_pybind/rocal_pybind.cpp:
```diff
@@ -583,6 +597,8 @@ PYBIND11_MODULE(rocal_pybind, m) {
         return std::make_pair(labels_array, bboxes_array);
     });
+    m.def("getOneHotEncodedLabels", &wrapper_one_hot_label_copy, py::return_value_policy::reference);
+    m.def("getCupyOneHotEncodedLabels", &wrapper_cupy_one_hot_label_copy, py::return_value_policy::reference);
```
Remove cupy and use gpu instead

swetha097 replied:
this is explicity for handling cupy arrays. So letting this be as is

sampath1117 commented on rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:
```diff
@@ -48,7 +48,7 @@ def main():
     device = "cpu" if rocal_cpu else "gpu"
     batch_size = args.batch_size
-    one_hot_labels = 0
+    one_hot_labels = 1
```
Why this has been changed to 1?

sampath1117 commented on rocAL_pybind/rocal_pybind.cpp:
```diff
@@ -129,6 +129,20 @@ py::object wrapperRocalExternalSourceFeedInput(
+    }
+
+    py::object wrapper_cupy_one_hot_label_copy(RocalContext context, size_t array_ptr, unsigned num_of_classes) {
+        void * ptr = (void*) array_ptr;
```
Please try to use static_cast instead of c style casting

sampath1117 commented on tests/cpp_api_tests/rocAL_unittests/rocAL_unittests.cpp:
```diff
@@ -689,38 +689,35 @@ int test(int test_case, int reader_type, const char *path, const char *outName,
                 char img_name[img_size];
-                numOfClasses = num_of_classes;
-                int label_one_hot_encoded[inputBatchSize * numOfClasses];
+                int label_one_hot_encoded[inputBatchSize * num_of_classes];
```
There seems to be mix and match of camelCase and snake case variable styling here. Please try to make all the variables style consistent

sampath1117 commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -255,18 +257,16 @@ void
     for (uint i = 0; i < meta_data_batch_size; i++) {
         int label_index = labels_buf[i];
-        if (label_index > 0 && label_index <= numOfClasses) {
+        if (label_index > 0 && label_index <= numOfClasses)
```
Please if you can change the numOfClasses to num_of_classes to make the style consistent with other variables

sampath1117 commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -255,18 +257,16 @@ void
             one_hot_encoded[(i * numOfClasses) + label_index - 1] = 1;
-
-        } else if (label_index == 0) {
+        else if (label_index == 0)
```
Use (!label_index) instead of (label_index == 0)

sampath1117 reviewed (commented):
Please address review comments

shobana-mcw commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -239,11 +239,13 @@ RocalTensorList
         THROW("Invalid rocal context passed to rocalGetOneHotImageLabels")
     auto context = static_cast<Context*>(p_context);
     auto meta_data = context->master_graph->meta_data();
+    auto meta_data_buffers = context->master_graph->get_meta_read_buffer_at(0);
```
Will we always get at index 0?

swetha097 replied:
The labels buffer is stored in ring buffer at 0th location.

shobana-mcw replied:
Then do you need an API with get_meta_read_buffer_at()?

If its always going to be labels the API name needs to be changed.

swetha097 replied:
this get_meta_read_buffer_at() can get from any location. But currently for this application , using 0 as the index to get labels

swetha097 replied:
removed the API as its not needed

shobana-mcw commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -239,11 +239,13 @@ RocalTensorList
     auto context = static_cast<Context*>(p_context);
     auto meta_data = context->master_graph->meta_data();
+    auto meta_data_buffers = context->master_graph->get_meta_read_buffer_at(0);
+    auto labels = meta_data.second->get_labels_batch();
```
Are you using this lables?mPlease check. 
What's the difference between meta_data_buffers and labels?

shobana-mcw commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -253,28 +255,26 @@ void
         THROW("meta data batch size is wrong " + TOSTR(meta_data_batch_size) + " != " + TOSTR(context->user_batch_size()))
 
     int labels_buf[meta_data_batch_size];
```
Why are you meta_data_batch_size by calling meta_data.second->get_labels_batch().size() when labels has already been fetched previously?

If labels is not used rempve that variable.

swetha097 replied:
Removed it

shobana-mcw commented on rocAL_pybind/examples/PYTHON_UNITTEST_TEST_FILE.sh:
```diff
@@ -23,10 +23,10 @@ golden_output_path=${ROCAL_DATA_PATH}/rocal_data/GoldenOutputsTensor/
 display=0
 batch_size=2
-device=0
+device=1
```
Default will be host ryt? Why are we changing to GPU?

shobana-mcw commented on rocAL_pybind/examples/rocAL_api_tf_classification_reader.py:
```diff
@@ -48,7 +48,7 @@ def main():
     batch_size = args.batch_size
-    one_hot_labels = 0
+    one_hot_labels = 1
     num_threads = args.num_threads
```
DO we have a readme for this example? 
I want us to mention that the one hot endocing  use case is demonstrated here.

swetha097 replied:
No @shobana-mcw - there is not explicit README for this particular file alone

swetha097 replied:
No, its not a part of the general README too

shobana-mcw reviewed (commented):
Does our one hot encoding support pytorch framework / GPU buffers?
Looks like this is specific to cupy buffers.
Please address the comments.

swetha097 commented:
> Does our one hot encoding support pytorch framework / GPU buffers? Looks like this is specific to cupy buffers. Please address the comments.

Yes, One hot encoding supports the torch buffers too

fiona-gladwin commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -253,28 +254,26 @@ void
     int labels_buf[meta_data_batch_size];
-    int one_hot_encoded[meta_data_batch_size * numOfClasses];
-    memset(one_hot_encoded, 0, sizeof(int) * meta_data_batch_size * numOfClasses);
-    memcpy(labels_buf, meta_data.second->get_labels_batch().data(), sizeof(int) * meta_data_batch_size);
```
@swetha097 Why can't we retain this way of getting labels, This line should work fine

Please check if this works, that way we need not introduce the get_meta_read_buffer_at API
Since the labels data is present in the metadata batch and in ring buffer

swetha097 replied:
Tested Fiona, getting all 0s if its used as earlier

swetha097 replied:
This seems to give all 0s as outputs

swetha097 replied:
done

fiona-gladwin commented:
# Get void pointer from NumPy array
void_ptr_numpy = arr_numpy.ctypes.data

# Get void pointer from CuPy array
void_ptr_cupy = arr_cupy.data.ptr

These can be done in tf.py itself and we can pass pointer directly right? Will that be acceptable

SundarRajan28 committed:
Adding support for file list reader

sampath1117 reviewed (commented):
Please address review comments

SundarRajan28 committed:
Merge branch 'swbs/audio/pr2' into swbs/audio/pr4

SundarRajan28 committed:
Adding file list reader to audio python example

SundarRajan28 committed:
Fixing reader import issue

SundarRajan28 committed:
Updating python example

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Fixing issues with file list reader

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Merge branch 'swbs/audio/pr2' into swbs/audio/pr3

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Removing storage_type from audio loader API

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr2' into audio_pr4

SundarRajan28 committed:
Initial commit for removing file list reader

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Removing file list reader and metadata reader

SundarRajan28 committed:
Merge branch 'swbs/audio/pr2' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Changing python unittests for QA mode

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Adding comment for file list case in file source reader

SundarRajan28 committed:
Merge branch 'swbs/audio/pr1' into swbs/audio/pr3

SundarRajan28 committed:
Merge branch 'swbs/audio/pr2' into swbs/audio/pr3

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Resolving review comments

shobana-mcw reviewed (commented):
Please address the comments.

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Adding file list reader to C++ unit tests

SundarRajan28 committed:
Fixing issues with C++ audio unit tests

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Updating audio unit tests for default file list path

fiona-gladwin committed:
Minor changes

shobana-mcw reviewed (commented):
Please address the review comments.

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Update CHANGELOG.md

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

kiritigowda committed:
Readme - Updates (#153)

* Readme - Updates

* Setup - Cleanup

* CMakeLists - remove OBSOLETES

* Readme - updates

* Updates

swetha097 committed:
Merge branch 'swbs/audio/pr3' into audio_pr4

SundarRajan28 committed:
Resolving review comments

SundarRajan28 committed:
Removing file_list_path from label_reader

samjwu committed:
Docs - Update Read the Docs configuration to use Python 3.10 and latest rocm-docs-core (#158)

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.2.0 to 1.2.1 in /docs/sphinx (#159)

Bumps [rocm-docs-core[api_reference]](https://github.com/RadeonOpenCompute/rocm-docs-core) from 1.2.0 to 1.2.1.
- [Release notes](https://github.com/RadeonOpenCompute/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/RadeonOpenCompute/rocm-docs-core/compare/v1.2.0...v1.2.1)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

LakshmiKumar23 committed:
Audio - Version check  (#156)

* version check from MIVisionX vx_rpp for Audio

* check if version file exists

* check for vx_rpp version and sndfile before settign rocal_audio

* remove unnecessary find_pacakge

* check for sndfile only if vx_rpp supports audio

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into audio_pr4

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.2.1 to 1.3.0 in /docs/sphinx (#163)

Bumps [rocm-docs-core[api_reference]](https://github.com/RadeonOpenCompute/rocm-docs-core) from 1.2.1 to 1.3.0.
- [Release notes](https://github.com/RadeonOpenCompute/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/RadeonOpenCompute/rocm-docs-core/compare/v1.2.1...v1.3.0)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kiritigowda committed:
Setup - Updates (#164)

* Setup - Updates

* Setup - pip packages updates

* CMakeList - Add Protobuf deps

swetha097 committed:
Audio PR - Augmentation support [PreEmphasis Filter] (#122)

* Audio Decoder PR 1

* channge image_info to sample_info to maintain a generic name for all the use-cases

* Change the copyright year from 2023 to 2024

* formatting the files

* Resolve PR comments

* NWC

* Resolve PR comments

* Change decoded_img_info to decoded_video_info

* Change the file_path() function to virtual from pure virtual

* Minor change

* Minor changes

* Add the unit test file

* Revert "Add the unit test file"

This reverts commit e79cc06d3c9d167dc8d142d315b71a5bd5e86736.

* Fix the ROI api in python end

* Fix the ROI api in python end

* Remove the ROI change from pytorch.py

* Add Python Unit Test for Audio

* PreEmphasis Filter

* Add Preemp filter augmentations files

* Introduce CMake for sndfile

Modify CMakeLists.txt for the same

* Resolve 1st set of PR commenst

* Remove commented code for last batch polices and unsued imports

* ROI related changes - change from xy to wh to use for samples and channels

* Fix seg fault with ROI

* Remove opencv usage from the unit test

* Resolve the PR comments

* Remove instances of the audio_*_time - use the existing variables from Timing struct

* Formatting changes in rocal_api_data_loader.cpp and add the opencl and hip conditions for audio loaders

* Resolve the internal PR comments

* Reformatting the file_source_reader.cpp

* Remove _input_path from audio_source_evaluator and audio_read_and_decode as it is unecessary

* Change the header formatting

* Changes in copy_data() for audio samples

* Initialize the status at the beginning

* Resolve PR comments

* Resolving PR comments

* Resolve PR comments

* Remove the unused variables

* Resolve PR commensts

* Cmake related changes for audio

* Resolve PR comments

* Add condition check to eliminate any other file extensions other than a wav file / other image formats and call open_folder deom subfolder_reading() function

* Add basic testing for Audio to compare Bin files

* Update audio_read_and_decode.cpp

* Update rocAL_api_audio_unittest.py

* Revert file source reader changes

* Update master_graph.cpp

* Update tensor.cpp - Remove a commented line of code

* Rephrase

* Introduce ROCAL_AUDIO flag

Introduce flag for audio code, to be disabled when sndfile not found

* Minor changes

* Minor changes

* Add output comparision for Audio outputs

* Minor changes

* Add comment

* Minor changes to unit test

* Remove max_frames and max_channels args

* Remove max_frames, max_channels and sample rate from unit test

* Remove the NONE types

* Clean up unit test

* Clean up code in audio python unit test

* Introduce ROI in the unit test

* Minor change

* Format python unit test

* Add python script to run audio unittests

* Fix formatting issues - Minor changes

* Clean up C++ audio unit test

* Remove max frames and channels from decoder

* Minor changes

* Add Output comparison for python audio unittests

* Modify rocal audio unit test

Update README

* Minor change

* Resolve PR comments

* Minor changes - Modifying the names of the arguments

* Added a bried desc for the rocAL enum for border type

* Add a WRN statement in PreEmphasis Filter to only use FP32 dtype

* minor changes

* Change the borderType enum to int32 from uint32 dtype

* Minor changes

* Minor change

* Update README for audio unit test

* Parameters for rocALAudioIterator

* Minor change

* Remove the reset_tensor_roi() from the PreEmphasis augmentation making sure the src and dst roi points to the same location

* Del the Unit Test Files introduced earlier

* Add pre_emphasis function and gollden output comparison in audio unit tests

* minor change - add update val in create array

* Minor variable name change

* minor change

* Update unit test

* Minor change

* minor change

* Minor changes

Add wav extension in file reader
Add reader in unit test

* Minor change

* Add file reader to python audio unit test

* Update C++ unit test

* Update python audio unit test

* Remove the unused variable output - Resolve warnings in cpp unit test

* Remove the dst_roi arg passed to rpp

* Minor changes

Change borderType enum prefix

* Name change from sample to data

* Change from decoded_data_info to DecodedDataInfo

* Remove audio_decoder_factory.cpp file

* Minor change

* Change variable name

* Update the struct variable name in audio files

* Minor changes

* Change ROCAL_DATA_PATH to exclude rocal_data

* Update ROCAL_DATA_PATH to exclude rocal_data

* Use Pascal case for function names in audio decoder

* Modify cmake to have SNDFILE in all capital

* Minor changes

* Add struct for audio info in AudioReadAndDecode

* Fix merge conflict

* Renaming crop_image_info to CropImageInfo

* Remove - actual_host_buffers - Unused

* Rename TimingDBG to TimingDbg

* Move the instances of DecodedDataInfo to its base class LoaderModule

* Fix a WRN msg in master_graph.cpp

* Remove a dangling comment

* Rename _circ_data_info to _circ_buff_data_info

* Add Glob to CMakeLists.txt

* Rename SndFileDecoder to GenericAudioDecoder

* Fix build issues

* Minor change

* Update python API README.md for audio unit test

* Update audio unit test README

* Revert "Add Glob to CMakeLists.txt"

This reverts commit 47263d94393e9614a5cebcab6c11c804ac7152f8.

* Fix include headers for Audio files

* Fix copy data 2D

* Minor changes

* Pass decoded data info to load routine instead of separate vectors

* Update CHANGELOG.md

* Update CHANGELOG.md

* Change swap_handle_time variable name in loader

* Update the changelog.md

* Formatting changes

Add comments

* Update doxygen comments

* Move file source reader from readers/image to readers folder

* Update README and add doxygen description

* Update CMakeLists and README for audio test

* Update README for audio test

* Minor fix

* Fix merge from PR 2

* Minor changes shard_count argument name

* Rename set and get functions of data_info to decoded_data_info

* Revert empty line removed in CMakeLists.txt

* Removed prefix original for audio vectors

* Resolve PR comments

* Add @params to all args in pytorch.py

* Minor changes in unit test

* Minor changes

* Change ROCAL instaces to rocAL in pytorch.py

* Resolve the PR comments

* Minor changes in decoders.py - Modify the comment for shard_size

* Minor changes

* Address the PR comments

* Address Review comments

---------

Co-authored-by: root <root@ixt-sjc2-54.local.lan>
Co-authored-by: Swetha B S <swetha@multicorwareinc.com>
Co-authored-by: fiona-gladwin <fionagladwin@multicorewareinc.com>
Co-authored-by: Fiona-MCW <70996026+fiona-gladwin@users.noreply.github.com>
Co-authored-by: Swetha B S <>
Co-authored-by: SundarRajan28 <sundarrajan@multicorewareinc.com>
Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

kiritigowda committed:
Readme - Updates (#165)

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into audio_pr4

dependabot[bot] committed:
Bump rocm-docs-core[api_reference] from 1.3.0 to 1.4.0 in /docs/sphinx (#166)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.3.0 to 1.4.0.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.3.0...v1.4.0)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

SundarRajan28 committed:
Merge remote-tracking branch 'upstream/develop' into audio_pr4
